packages/sdk/src/evm/core/classes/staking-721.ts

```diff
@@ -0,0 +1,258 @@
+import { NetworkInput } from "..";
+import {
+  fetchCurrencyMetadata,
+  fetchCurrencyValue,
+} from "../../common/currency";
+import { resolveAddress } from "../../common/ens";
+import { buildTransactionFunction } from "../../common/transactions";
+import {
+  FEATURE_NFT_STAKE,
+  FEATURE_TOKEN_STAKE,
+} from "../../constants/thirdweb-features";
+import { Address, AddressOrEns } from "../../schema";
+import { Currency, CurrencyValue, Amount } from "../../types";
+import { BaseERC20, BaseERC721 } from "../../types/eips";
+import { DetectableFeature } from "../interfaces/DetectableFeature";
+import { UpdateableNetwork } from "../interfaces/contract";
+import { ContractWrapper } from "./contract-wrapper";
+import { Transaction } from "./transactions";
+import type {
+  TokenStake,
+  Staking20Base,
+  NFTStake,
+  Staking721Base,
+} from "@thirdweb-dev/contracts-js";
+import { ThirdwebStorage } from "@thirdweb-dev/storage";
+import ERC20Abi from "@thirdweb-dev/contracts-js/dist/abis/IERC20.json";
+import ERC721Abi from "@thirdweb-dev/contracts-js/dist/abis/IERC721.json";
+import { BigNumber, BigNumberish, ethers } from "ethers";
+import { AmountSchema } from "../../../core/schema/shared";
+
+/**
+ * Standard ERC20 Token functions
+ * @remarks Basic functionality for a ERC20 contract that handles all unit transformation for you.
+ * @example
+ * ```javascript
+ * const contract = await sdk.getContract("{{contract_address}}");
+ * await contract.erc20.transfer(walletAddress, amount);
+ * ```
+ * @public
+ */
+export class Staking721<T extends NFTStake | Staking721Base>
+  implements UpdateableNetwork, DetectableFeature
+{
+  featureName = FEATURE_NFT_STAKE.name;
+  protected contractWrapper: ContractWrapper<T>;
+  protected storage: ThirdwebStorage;
+  protected _rewardToken: ContractWrapper<BaseERC20> | undefined;
+  protected _stakingToken: ContractWrapper<BaseERC721> | undefined;
+
+  private _chainId: number;
+  get chainId() {
+    return this._chainId;
+  }
+
+  constructor(
+    contractWrapper: ContractWrapper<T>,
+    storage: ThirdwebStorage,
+    chainId: number,
+  ) {
+    this.contractWrapper = contractWrapper;
+    this.storage = storage;
+    this._chainId = chainId;
+
+    this.getContractWrappers();
+  }
+
+  /**
+   * @internal
+   */
+  onNetworkUpdated(network: NetworkInput): void {
+    this.contractWrapper.updateSignerOrProvider(network);
+  }
+
+  /**
+   * @internal
+   */
+  getAddress(): Address {
+    return this.contractWrapper.readContract.address;
+  }
+
+  ////// Standard Staking721 Extension //////
+
+  // READ FUNCTIONS
+  public async getRewardTokenBalance(): Promise<CurrencyValue> {
+    return fetchCurrencyValue(
+      await this.contractWrapper.getProvider(),
+      (await this.getRewardToken()).address,
+      await this.contractWrapper.readContract.getRewardTokenBalance(),
+    );
+  }
+
+  public async getRewardToken(): Promise<Currency & { address: string }> {
```

Comment:
would make sense to put this code in a common helper file or a base class. There's a bunch of common logic between the diff staking contracts

Reply:
One problem is that some functions might be similar but the underlying logic does differ in little ways. I'll do the moving to a base class once I'm finished with features. Then I'll find a way to make the classes have as less duplicate functions as possible